### DISCUSSION TITLE
Building TGN with Seastar

### DISCUSSION THREAD

**Discussion by JoelMathewC** [bfGraph/STGraph#64] [Ideas]
This is an attempt to build TGN with Seastar. 
Ref Paper: https://arxiv.org/abs/2006.10637

**Comment by JoelMathewC** [comment]
### Approach 1 (TCSR)
From the research I have done. Seastar might be useful in calculating embedding in TGN. In the GraphEmbedding  (TGN-sum and TGN-attn) models. A solid approach to implementing this would be to use the T-CSR datastructure detailed [here](https://arxiv.org/abs/2203.14883) and modify the seastar kernel to use this datastructure and process embeddings. 

### Approach 2 (GPMA)
However using a datastructure like GPMA might some benifits. For example in calculating the n-hop stuff. Additionally we can make it so that every batch in inserted into the GPMA and at every insertion we run a kernel that does a cleanup of old edges. We will probable need to modify GPMA to enable storage of timestamp along with the edges.

**Comment by JoelMathewC** [comment]
## TGN Checklist

- [x] For `backprop_every=1` Print out the time taken to do one forward prop and backward prop. Check what is the proportion of time taken by the embedding module alone.
~Make TGN such that all the inputs are GPU tensors, to avoid intermediate data transfers.~
~Verify why the GPU tensor based TGN is slower and calculate proportion of time taken by embedding in comparison to total time.~
- [ ] Print out the input and output shapes of embedding module
- [ ] Run seperate dummy seastar program to verify if we can replicate the same outputs for the given inputs
- [ ] Modify seastar template to process only the nodes in question for a particular batch. Verify if the time will be lesser than that taken by tgn's original embedding module.
- [ ] Modify GPMA to handle the insertion of nodes (with timestamps)
- [ ] Modify GPMA to remove neighbours that are no longer relevant
- [ ] Modify GPMA to handle having to backpropagate after certain number of batches rather than having to backpropagate after the whole thing. (As a starting we can set `backprop_every=1` and simply used count sort to build the reverse graph for each time)
- [ ] Copy tgn into the nn/ folder in seastar
- [ ] Copy the train script to benchmarking/ and verify if it runs
- [ ] Modify train script to insert to GPMA
- [ ] Modify nn/tgn to use seastar embedding module
- [ ] Put a `NotImplementedError` in backward to verify that its being triggered.
- [ ] Verify if both TGNs achieve same accuracy, verify the amount of time taken.

**Reply by JoelMathewC** [reply]
## Benefits of using Seastar

1. When implementing more complex message functions, we will need access to the a nodes neighbours. Seastar's GPMA graph object will give quick access to these neighbours. Additionally, Seastar will provide an easy way to implement (vertex-centric program) and fast execution of message function.
2. Graph based embeddings are crucial for performance of TGN. Seastar can help speed that up and also make it easier to write the logic (vertex-centric program)

**Comment by JoelMathewC** [comment]
## 🤔 What benefit can Seastar provide to TGN?

We are going to verify what proportion of time is spend in calculating the graph embeddings to understand if Seastar can provide a considerable improvement that is worth implementing.

### Measuring time
(Original TGN) These are the results of measuring the time taken to do one forward and backward propagation along with the time taken to compute embeddings from two different embedding modules.

```
GRAPH_SUM
-------------
Time taken for Compute Embedding: 0.7638850212097168
Batch 0  | Forward Prop Time: 0.8806066513061523
After Backprop | Total time taken: 0.890772819519043

Time taken for Compute Embedding: 0.008257150650024414
Batch 0  | Forward Prop Time: 0.024766921997070312
After Backprop | Total time taken: 0.04060697555541992

Time taken for Compute Embedding: 0.007761478424072266
Batch 0  | Forward Prop Time: 0.022415637969970703
After Backprop | Total time taken: 0.035907745361328125

Time taken for Compute Embedding: 0.007882118225097656
Batch 0  | Forward Prop Time: 0.02371358871459961
After Backprop | Total time taken: 0.03816032409667969

Time taken for Compute Embedding: 0.007735013961791992
Batch 0  | Forward Prop Time: 0.023021221160888672
After Backprop | Total time taken: 0.037084341049194336

Time taken for Compute Embedding: 0.008192062377929688
Batch 0  | Forward Prop Time: 0.02449941635131836
After Backprop | Total time taken: 0.03934526443481445
```

```
GRAPH_ATTN
--------------

Time taken for Compute Embedding: 0.5725018978118896
Batch 0  | Forward Prop Time: 0.6631417274475098
After Backprop | Total time taken: 0.6741161346435547

Time taken for Compute Embedding: 0.0054166316986083984
Batch 1  | Forward Prop Time: 0.01765584945678711
After Backprop | Total time taken: 0.033254384994506836

Time taken for Compute Embedding: 0.00520014762878418
Batch 2  | Forward Prop Time: 0.01636648178100586
After Backprop | Total time taken: 0.03276205062866211

Time taken for Compute Embedding: 0.0055332183837890625
Batch 3  | Forward Prop Time: 0.0170900821685791
After Backprop | Total time taken: 0.034546613693237305

Time taken for Compute Embedding: 0.006359577178955078
Batch 4  | Forward Prop Time: 0.02056717872619629
After Backprop | Total time taken: 0.0368196964263916

Time taken for Compute Embedding: 0.006069660186767578
Batch 5  | Forward Prop Time: 0.018717050552368164
After Backprop | Total time taken: 0.03549933433532715
```


Proportion of time for embedding calculation in `graph_sum` : 33%
Proportion of time for embedding calculation in `graph_attn` : 31%

**Comment by JoelMathewC** [comment]
## Seastar-TGN Design

### GPMA design

GPMA can be designed such that the `keys` are _(src, timestamp, dst)_, this will ensure that for each node its k-most recent neighbours are easily accessible. (Most recent sampling)

For backpropagation, if it happens after every batch then we simply have to generate the reverse CSR with our `count sort` based algorithm. But if it is happening after a few batches then we will have to generate a deep copy of the GPMA after a few batches and then progressively delete the updates and generate the reverse CSR and perform backprop.

#### Handling uniform sampling
However, we are now left with the more challenging uniform sampling. There is a question of how data can be shuffled and how we will access the same data during backpropagation. One option is for every thread mapped to a node to generate a random list of number with the max value of a numbering generated being the maximum number of neighbours they have. They can generate k such numbers. The thread can then get the k-outputs, save them on a seperate list in a stack and use it. During backprop that stack is popped to get the result. We'll leave this idea here and not implement it and just proceed with what we had before.

Another option for uniform sampling is to build the reverse CSR for each batch and when building it making sure that the ones being sampled are marked and only those are being converted into the reverse CSR. These reverse CSRs can then be cached and executed only when backprop is triggered. If we are only building the CSR for the nodes being processed the memory should not be a concern.

### Seastar Design
We'll modify the Seastar frontend to accept an input of `source nodes` and `destination nodes`, the source nodes will be a param in the forward kernel and destination nodes a param in the backward kernel. We can save the destination nodes on the `state-stack` so that they are accessible when they are required.

The Raw-message store, (old) messages, aggregated messages, memory and embeddings are all aspects of a single node and hence fit into the vertex-centric scheme. It should be possible to model a vertex centric program that performs all these operations without materializing internal results. Additionally it would make the code more easier.

Also we will need to modify the kernel to only operate on the nodes that are in question and also only on the neighbours that we are considering at that time. This may need some modifications in the frontend.

Additionally, the biggest challenge we face is that of using linear layers on concatenated results, I do not know how we will be able to work with that. We might need to handle those externally.

**Reply by JoelMathewC** [reply]
## Sample

A sample of how the `TGN` code will look. It may be noted that Seastar creating return tensors gives us the advantage of generating a node feature matrix of node's in question in a particular iteration. Note the following
1. The final stage of adding the raw updates is just a GPMA update and doesnt need to be done in the vertex centric part.
2. The final Graph sum can have multiple layers so that can be abstracted into a seperate class. Actually everything can be seperate classes.

```
import torch
import torch.nn as nn
from seastar.compiler import Seastar
from seastar.compiler.backend.pytorch.torch_callback import SeastarBackendTorch


# Assumptions
# AGGREGATION = LATEST MESSAGE
# NODE EMBEDDING = GRAPH SUM

class TGN(nn.Module):
    def __init__(self):
        super(TGN, self).__init__()
        
        self.gru = nn.GRUCell(input_size=message_dimension,
                              hidden_size=memory_dimension)
        self.w1 = nn.Parameter()
        self.w2 = nn.Parameter()

        self.seastar = Seastar(SeastarBackendTorch())
        self.reset_parameters()

    def reset_parameters(self):
        pass

    def forward(self, g, h, edge_weight=None):

        # Raw messages and message (since this is identity)
        @self.seastar.compile(gnn_module=self)
        def get_raw_messages(v):
          latest_in_edge = v.inedges[0]
          latest_out_edge = v.outedges[0]
          if latest_out_edge.time > latest_in_edge.time:
            return v.memory, latest_edge.dst.memory, latest_edge.time, latest_edge.feat
          else:
            return v.memory, latest_edge.src.memory, latest_edge.time, latest_edge.feat

        raw_msgs = nb_compute(g=g, n_feats={'memory': None}, e_feats={'time': None, 'feat': None}, nodes=None) # nodes is a list of nodes involved in this batch src/dst

        # Aggregated Message
        old_msgs = torch.cat([raw_msgs[0], raw_msgs[1], raw_msgs[2], raw_msgs[3]])
        old_mem = raw_msgs[0] # this has old memory with shape (len(nodes), memory_dim) 

        # Updated Memory
        updated_mem = self.gru(old_msgs, old_mem)


        # Node embeddings
        h1 = None
        for i in range(num_layers):
          @self.seastar.compile(gnn_module=self)
          def get_node_emb(v):
            h = sum([ torch.mm(self.w1, torch.cat([edge.dst.feat + edge.dst.memory, edge.feat, edge.time - v.latest_time])) for edge in v.get_outedges(n_neighbours)])
            h = self.relu(h)
            return h
          h = nb_compute(g=g, n_feats={'memory': None}, e_feats={'time': None, 'feat': None}, nodes=None) # nodes is a list of nodes involved in this batch src/dst

          if h1 is not None:
            h1 = torch.mm(self.w2, torch.cat([h1,h]))
          else:
            h1 = h
        
        return h1
```

**Reply by JoelMathewC** [reply]
### Challenges

The major challenges with this approach is
1. Need for temporal graphs to access both in and out edges. Either we have two GPMAs or we have one GPMA and construct the opposite GPMA each time. I'm not sure on how that may affect the speed.
2. Very complicated node embedding operation requiring concatenation operations and matrix multiplications followed by the sum. We need a smart way to move that logic outside to make this easier.
3. There is no support for edge indexing on seastar to get nearest edge
4. No support to access outedges in seastar

**Reply by JoelMathewC** [reply]
### An idea to solve challenge 1

This might be a slightly risky idea, but I was wondering if it would be possible to use GPMA as a base and then construct forward and reverse CSR for the graph in question. That would mean all source nodes, destination nodes and their 1-hop sampled neighbours. In the case we are considering higher hops we can increase the reach and maybe bfs could help us out. Also both the CSRs can be generated at the same time, with a thread constructing a part of both CSRs

1. Constructing of forward and reverse CSRs can be done with the algorithm we developed in #56. 
2. This would also mean that we wouldn't have to edit the seastar kernel and we can use it as is.

The reverse CSRs can be stored till backprop happen but they also serve the purpose of giving us access to the out edges which has become important.

If this idea is used then the n_neighbours can be specified to GPMA before hand and we would not need to edit Seastar for that.

**Reply by JoelMathewC** [reply]
### Addressing challenge 2

Seastar does shape propagation, so it should be able to predict the longer output we expect. We just need a implementation for concatenation in Seastar. If we can return that output then we could perform the matrix multiplication outside and then do the sum outside as well. But I feel like that could impact the backprop.

We would additionally then have to figure out how to do matrix multiplication inside Seastar, unless its already been done. We will have to verify that tomorrow.

- [ ] Verify if `matmul` is already implemented in seastar
- [ ] Look for ways to implement concatenation in seastar.

### A new approach
The node features and edge features in question can be obtained from `torch.select_index`. Basically the theoretical idea is if `a` are node features and `b` are edge features. The run them both through seperate linear layers. Then use seastar's vertex centric aggregation to obtain on the result, but when doing it make sure you add the the result of processing `a` to the result of `b` during aggregation. Hence no need for concat or linear layer. (need to verify if backpropagation works though).

**Reply by JoelMathewC** [reply]
### Addressing challenge 4

For simplicity we can consider the case of undirected graph as the original TGN does. In this case you will also need to generate just one CSR and that would be the backward CSR as well.

At any point when we do implement a directed graph we can use the forward and reverse CSRs to obtain the in- and out-edges.

**Reply by JoelMathewC** [reply]
### Updated Design

#### Constraints
1. All edges are undirected
2. We're considering most recent message for message aggregation.
3. We're considering k-most recent nodes for sampling
4. We're considering graph sum for node embedding calculation

#### Details
1. GPMA is modified to store `(src,timestamp,dst)` as keys. Additionally we will be storing proper eids as values since each edge has a unique eid. (Remember that eid != 0 and eid != 1).
7. Before every forward pass we use the count sort logic to sample k-most recent neighbours from each node and then generate a forward and backward CSR with only the nodes relevant to that pass.
8. We maintain tensors: node_memory, node_last_msg_as_src, node_last_msg_as_dst, node_features, edge_features, edge_timestamps.
9. At a given forward pass we accept as input: src_nodes, dst_nodes, neg_nodes, edge_times
10. At the start of each forward pass we will use `torch.select_index` to filter all the tensors we maintain to hold only the ones relevant to that pass.
11. The messages to be used are already present in node_last_msg_as_src, node_last_msg_as_dst
12. **Aggregation Step:** We can use a normal seastar kernel with node_last_msg_as_src, node_last_msg_as_dst passed as `n_feats` to aggregate the messages. For identity functions we can use the messages directly from the tensors.
13. **Memory Step:** Pass the memory and aggregated messages through an RNN layer
14. **Node Embedding:** Use temporally-aware seastar that accepts a list of [src_nodes, dst_nodes, neg_nodes] along with [edge_times, edge_times, edge_times] to generate node embeddings for each node for the different timestamps corresponding to those nodes. The output will be of format [len(src_nodes + dst_nodes + neg_nodes), node_emb_dim]. Also we will multiple edge features with a weight matrix before the vertex centric program, similarly for the previous node embedding and timestamp feature of each edge and the timestamp feature of the edge times list mentioned above. We will return the aggregated edge_feat and timestamp results. These two will then be concatenated with the previous node embedding.
15. **Calculate Loss/ Finish Training**
16. **Computing messages** Compute messages for the interactions of this batch and store them in the respective node_last_msg_as_src, node_last_msg_as_dst
17. **Updating memory** Store the updated memory of the nodes in the current forward pass back in the node_memory tensor.

**Reply by JoelMathewC** [reply]
### Temporally Aware Seastar

**Input:** list of nodes, list of timestamp features, n_feats, e_feats
Note: list of nodes and list of timestamps have the same 0th dimension

#### Modifications
1. the output tensor will be of shape [len(src_nodes + dst_nodes + neg_nodes), node_emb_dim]
2. The kernel will be such that one group is assigned to each element on the list of nodes. This means multiple groups could be mapped to the same node just at different timestamps.
3. The timestamp that the current node is being considered for can be accessed in the frontend. Instead of time stamp features we could have a t_feats. We can think of that.

**Reply by JoelMathewC** [reply]
### Bypassing concatenation and weight multiplication
Both the following code segments produce the same result on training

```
class Model1(nn.Module):
  def __init__(self, n_input, n_output):
    super(Model1, self).__init__()
    self.linear = nn.Linear(n_input, n_output)
  
  def forward(self,feat):
    return self.linear(feat)
```

```
class Model2(nn.Module):
  def __init__(self, n1_input, n2_input, n_output):
    super(Model2, self).__init__()
    self.linear1 = nn.Linear(n1_input, n_output)
    self.linear2 = nn.Linear(n2_input, n_output)
  
  def forward(self,feat1, feat2):
    return self.linear1(feat1) + self.linear2(feat2)
```
